IdentityCore/tests/integration/ios/MSIDDefaultBrokerResponseHandlerTests.m

```diff
@@ -0,0 +1,488 @@
+    XCTAssertEqual([accessTokens count], 1);
+    
+    MSIDAccessToken *accessToken = accessTokens[0];
+    XCTAssertEqualObjects(accessToken.accessToken, @"i-am-a-access-token");
+    XCTAssertEqualObjects(accessToken.scopes, [scopes msidScopeSet]);
+    XCTAssertEqualObjects(accessToken.clientId, @"my_client_id");
+    XCTAssertEqualObjects(accessToken.authority.url.absoluteString, @"https://login.microsoftonline.com/contoso.com-guid");
+    XCTAssertEqualObjects(accessToken.expiresOn, result.accessToken.expiresOn);
+    XCTAssertEqualObjects(accessToken.extendedExpireTime, result.accessToken.extendedExpireTime);
+    XCTAssertEqualObjects(accessToken.accountIdentifier.homeAccountId, @"1.1234-5678-90abcdefg");
+    
+    //Check refresh token in cache
+    NSArray *refreshTokens = [MSIDKeychainTokenCache getAllDefaultRefreshTokens:self.cacheAccessor];
+    XCTAssertEqual([refreshTokens count], 2);
+    
+    MSIDRefreshToken *refreshToken = refreshTokens[0];
+    XCTAssertEqualObjects(refreshToken.refreshToken, @"i-am-a-refresh-token");
+    XCTAssertEqualObjects(refreshToken.familyId, @"1");
+    XCTAssertEqualObjects(refreshToken.accountIdentifier.homeAccountId, @"1.1234-5678-90abcdefg");
+    XCTAssertEqualObjects(refreshToken.authority.url.absoluteString, @"https://login.microsoftonline.com/common");
+    
+    //Check id token in cache
+    NSArray *idTokens = [MSIDKeychainTokenCache getAllIdTokens:self.cacheAccessor];
+    XCTAssertEqual([idTokens count], 1);
+    
+    MSIDIdToken *idToken = idTokens[0];
+    XCTAssertEqualObjects(idToken.authority.url.absoluteString, @"https://login.microsoftonline.com/contoso.com-guid");
+    XCTAssertEqualObjects(idToken.rawIdToken, idTokenString);
+    
+    //Check account in cache
+    NSArray *accounts = [self.cacheAccessor allAccountsForAuthority:nil clientId:nil familyId:nil context:nil error:nil];
+    MSIDAccount *account = accounts[0];
+    XCTAssertEqualObjects(account.authority.url.absoluteString, @"https://login.microsoftonline.com/contoso.com-guid");
+    XCTAssertEqualObjects(account.username, @"user@contoso.com");
+    XCTAssertEqualObjects(account.accountIdentifier.homeAccountId, @"1.1234-5678-90abcdefg");
+    XCTAssertEqualObjects(account.accountIdentifier.legacyAccountId, @"user@contoso.com");
+    XCTAssertEqualObjects(account.clientInfo.rawClientInfo, rawClientInfo);
+}
+
+- (void)testHandleBrokerResponse_whenValidBrokerErrorResponse_shouldReturnNilResultAndError
+{
+    [self saveResumeStateWithAuthority:@"https://login.microsoftonline.com/common"];
+    
+    NSString *correlationId = [[NSUUID UUID] UUIDString];
+    
+    NSDictionary *errorMetadata =
+    @{
+      @"http_response_code" : @200,
+      @"error_description" : @"Error occured",
+      @"oauth_error" : @"invalid_grant",
+      @"oauth_sub_error" : @"consent_required",
+      @"declined_scopes" : @"decliendScope1 decliendScope2",
```

Comment:
I think declined scopes handling should happen on MSAL side, otherwise MSAL will never cache those tokens and next request will result in interactive auth. Can you make sure the protocol and code accounts for those cases?

Reply:
Yes, makes sense :)

Reply:
Discussed with Olga again, we decided to send declined_scopes as well as granted_scopes. Besides of that, we also send the tokens in the` additional_tokens` field.